### DISCUSSION TITLE
Deck Driver not Compiling

### DISCUSSION THREAD

**Discussion by tjt22** [bitcraze/discussions#306] [Q&A] [answered]
I'm trying to make a custom deck driver for a hall effect sensor that communicates over I2C and running into issues. The driver was heavily inspired by the examples on https://www.bitcraze.io/2020/07/intro-to-autonomous-robotics-with-the-crazyflie/ and https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/howto/. 

When I add the .o file in the Kbuild file in src/deck/drivers/src/ and do `make clean && make`, the file doesn't show up in the terminal like all the other Crazyflie deck drivers do. The same issue happens when following the hello.c tutorial on the Bitcraze website. 

Then after I flash it to the drone and check the console in the client, the following errors show:
`DECK_INFO: CONFIG_DECK_FORCE=whiskerDeck found`
`DECK_INFO: WARNING:  compile-time force driver whiskerDeck not found`
`DECK_CORE: 0 deck(s) found`

What am I doing wrong? Relevant images and code are below:

Kbuild file:
![image](https://user-images.githubusercontent.com/114026856/194143563-1f7bf8f4-286c-45e3-97ed-71e227d31c92.png)

Terminal after make clean && make with my driver not showing up:
![image](https://user-images.githubusercontent.com/114026856/194143428-6c016573-9acf-45eb-9dde-cc5d4599d7a0.png)

Driver Code: 

``` c
#define DEBUG_MODULE "whiskerDeck"

// Crazyflie system headers
#include "debug.h"
#include "i2cdev.h"
#include "deck.h"
#include "log.h"
#include "FreeRTOS.h"
#include "task.h"
#include "system.h"

// RTOS Stuff
#define WHISKER_TASK_STACKSIZE  (7*configMINIMAL_STACK_SIZE) 
#define WHISKER_TASK_NAME "WHISKER"
#define WHISKER_TASK_PRI 3

// variables
#define ADDRESS 0x35
#define DELAY 500
float x_flux;
float y_flux;
float z_flux;
static bool isInit;
void whiskerTask(void* arg);


static void whiskerInit()
{
    if (isInit)
        return;

    DEBUG_PRINT("Initializing Whisker...\n");
    i2cdevInit(I2C1_DEV);

    xTaskCreate(whiskerTask,WHISKER_TASK_NAME, WHISKER_TASK_STACKSIZE, NULL, WHISKER_TASK_PRI, NULL)
    
    isInit = true;
    DEBUG_PRINT("Whisker initialization complete!\n";)
}

static bool whiskerTest()
{
    DEBUG_PRINT("WhiskerDeck test\n");
    return true;
}

void whiskerTask(void* arg)
{
    // Read first 7 data bytes
    uint8_t buf[7]

    // Wait for system to start
    systemWaitStart();
    TickType_t xLastWakeTime;

    // UPDATE FOR MY SENSOR
    i2cdevWriteByte(I2C1_DEV, (uint8_t)ADDRESS, I2CDEV_NO_MEM_ADDR, (uint8_t)0b00100000);

    xLastWakeTime = xTaskGetTickCount();

    while (1) {
    vTaskDelayUntil(&xLastWakeTime, M2T(120));
   
    // Read seven bytes from the sensor
    i2cdevRead(I2C1_DEV, (uint8_t)BH1750I2CAddr, 7, &magdata);
    // Combine the seven bytes to get data
    int16_t x_flux = (int16_t)((magdata[0] << 8) | (magdata[4] & 0x0F)) >> 4;
    int16_t y_flux = (int16_t)((magdata[1] << 8) | ((magdata[4] & 0x0F) << 4)) >> 4;
    int16_t z_flux = (int16_t)((magdata[2] << 8) | ((magdata[5] & 0x0F) << 4)) >> 4;
    // DEBUG_PRINT("Light reading is: %f\n", intensity);
  }

}


// (rename source file pls)
static const DeckDriver WhiskerDriver = {
    .name = "mywhiskerDeck",
    .init = whiskerInit,
    .test = whiskerTest,
};



// Logging:
DECK_DRIVER(WhiskerDriver);
LOG_GROUP_START(Whisker)
LOG_ADD(LOG_FLOAT, x_flux, &x_flux)
LOG_ADD(LOG_FLOAT, y_flux, &y_flux)
LOG_ADD(LOG_FLOAT, z_flux, &z_flux)
LOG_GROUP_STOP(Whisker)
```

**Comment by ataffanel** [comment, marked as answer]
What you are doing seems correct. Adding the deck driver the way you are doing it does (try to) compile the driver for me. The driver does not compile so your problem is that the build system does not even try to build your driver.

Are you sure you have saved the `Kconfig` file? In your screenshot vscode shows that there is unsaved modification.

**Reply by tjt22** [reply]
Thank you for your response! I saved everything and had the same issue. Maybe I'm enabling the driver incorrectly. Here is what I did: 
- used the command `make menuconfig`
- navigated to `Expansion deck configuration`
- clicked on `Force load specified custom deck driver` and entered `mywhiskerDeck` which is the .name in my deck driver script
![image](https://user-images.githubusercontent.com/114026856/194347931-630ca313-722f-4d40-b489-57f9f3af37e9.png)
- then I clicked ok and saved the .config as .config
![image](https://user-images.githubusercontent.com/114026856/194347840-44e35619-df50-4934-9578-664ff23086e8.png)

I'm just saving everything to .config because I thought that's just the default config name. Is my process correct? Do I also need to modify the Kconfig file that's in src/deck/drivers/src/?

**Reply by tobbeanton** [reply]
Ahh, the menuconfig saving is not that intuitive. When you choose save config you actually "save as" a new config. You should instead press "exit" until you get the choice to save to current config.

**Reply by tjt22** [reply]
Ok so I did that and it says it saved to .config!
![image](https://user-images.githubusercontent.com/114026856/194543167-c427d9ca-635f-42ff-88e4-f5d2a9d7af92.png)

However, I still don't see my driver compiling when I `make clean && make`
![image](https://user-images.githubusercontent.com/114026856/194543107-42da29bd-62c2-4f67-934f-f6be73017b48.png)

**Reply by tobbeanton** [reply]
My bad, I thought you saved it to a new config with a different name.

It does not compile the whiskerDeck.c. Could there be some name mismatch problem or file in wrong folder? Try e.g. replace the activeMarkerDeck.o with whiskerDeck.o and see if it compiles then and not activeMarkerDeck.o.

**Reply by tjt22** [reply]
All good. So I tried that and replaced activeMarkerDeck in the kbuild file, but it still shows up after running `make clean && make` and still doesn't compile my script.

![image](https://user-images.githubusercontent.com/114026856/195078857-158ae3a4-1c7c-4a81-aabe-ff3590830e3e.png)

**Reply by ataffanel** [reply]
I feel we are stuck in a loop of "works for me" and not for you so I tried to start from the beginning: I installed the vm in Virtualbox 7.0 and try to document my steps so that you can repeat them and hopefully we get to the same result.

So, my understanding is that the main problem is that your driver is not compiling. I could get a file to compile with the following procedure:

 - Launch a clean Bitcraze vm
 - Open a terminal and type:
 - ```cd projects/crazyflie-firmware```
 - ```git pull```
 - ```make cf2_defconfig```
 - ```make```
 - ```touch src/deck/drivers/src/hello.c```  # Creates an empty C files in the deck folder
 - ```echo "obj-y += hello.o" >> src/deck/drivers/src/Kbuild```  # Adds the file to be always built
 - ```make```

This results in the new ```hello.c``` file to compile.

![image](https://user-images.githubusercontent.com/6381499/195304319-406a6ea9-deb8-4e9c-b465-198ad025f6d4.png)

If you follow the same procedure exactly you should get the file to compile and so you should get in a state where you can add your own files.

**Reply by tjt22** [reply]
It finally compiles! I think something was wrong with my VM. I updated to VirtualBox 7 and followed your instructions exactly. Thanks so much for the help. 

Now that it's compiling, I'm running into errors from my code. Do you mind helping me clear these? Do I have to clear these errors for the rest of the build to finish? 

This code is supposed to read the 3 cartesian components from a hall effect sensor. Based on this code: https://github.com/Infineon/TLI493D-W2BW/blob/master/examples/Raw_I2C_readout/Raw_I2C_readout.ino for this sensor: https://www.infineon.com/cms/en/product/evaluation-boards/s2go_3d_tli493dw2bw-a0/

![image](https://user-images.githubusercontent.com/114026856/198388516-b776e83f-2934-4871-832d-68719a9c5289.png)

``` c
#define DEBUG_MODULE "whiskerDeck"

// Crazyflie system headers
#include "debug.h"
#include "i2cdev.h"
#include "deck.h"
#include "log.h"
#include "FreeRTOS.h"
#include "task.h"
#include "system.h"

// RTOS Stuffcrazyflie
#define WHISKER_TASK_STACKSIZE  (7*configMINIMAL_STACK_SIZE) 
#define WHISKER_TASK_NAME "WHISKER"
#define WHISKER_TASK_PRI 3

// variables
#define ADDRESS 0x35
#define DELAY 500

static bool isInit;
void whiskerTask(void* arg);


static void whiskerInit()
{
    if (isInit)
        return;

    DEBUG_PRINT("Initializing Whisker...\n");
    i2cdevInit(I2C1_DEV);

    xTaskCreate(whiskerTask,WHISKER_TASK_NAME, WHISKER_TASK_STACKSIZE, NULL, WHISKER_TASK_PRI, NULL);
    
    isInit = true;
    DEBUG_PRINT("Whisker initialization complete!\n");
}

static bool whiskerTest()
{
    DEBUG_PRINT("WhiskerDeck test\n");
    return true;
}

void whiskerTask(void* arg)
{
    // Read first 7 data bytes
    uint8_t magdata[7];

    // Wait for system to start
    systemWaitStart();
    TickType_t xLastWakeTime;

    // UPDATE FOR MY SENSOR
    i2cdevWriteByte(I2C1_DEV, (uint8_t)ADDRESS, I2CDEV_NO_MEM_ADDR, (uint8_t)0b00100000);

    xLastWakeTime = xTaskGetTickCount();

    while (1) {
    vTaskDelayUntil(&xLastWakeTime, M2T(120));
   
    // Read seven bytes from the sensor
    i2cdevRead(I2C1_DEV, (uint8_t)ADDRESS, 7, &magdata);
    // Combine the seven bytes to get data
    int16_t x_flux = (int16_t)((magdata[0] << 8) | (magdata[4] & 0x0F)) >> 4;
    int16_t y_flux = (int16_t)((magdata[1] << 8) | ((magdata[4] & 0x0F) << 4)) >> 4;
    int16_t z_flux = (int16_t)((magdata[2] << 8) | ((magdata[5] & 0x0F) << 4)) >> 4;
    // DEBUG_PRINT("Light reading is: %f\n", intensity);
  }

}


// (rename source file pls)
static const DeckDriver WhiskerDriver = {
    .name = "mywhiskerDeck",
    .init = whiskerInit,
    .test = whiskerTest,
};



// Logging:
DECK_DRIVER(WhiskerDriver);
LOG_GROUP_START(Whisker)
LOG_ADD(LOG_FLOAT, x_flux, &x_flux)
LOG_ADD(LOG_FLOAT, y_flux, &y_flux)
LOG_ADD(LOG_FLOAT, z_flux, &z_flux)
LOG_GROUP_STOP(Whisker)
```

**Reply by ataffanel** [reply]
You should remove the ```&``` on ```&magdata```: in C an array name can be used directly to get a pointer to the first element. If you try to get a pointer to it you get a pointer to the pointer to the first element (very similar to a ```uint8_t **```).

Ps. I have been editing your messages with \`\`\`c  \<source code\> \`\`\`. This makes the C code much easier to read.

**Reply by tjt22** [reply]
Ahhh okay that fixed it. Thanks so much for the help, you guys are the best!
